src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1137 @@
+
+    fn process_get_pin_retries(
+        &self,
+        persistent_store: &PersistentStore,
+    ) -> Result<AuthenticatorClientPinResponse, Ctap2StatusCode> {
+        Ok(AuthenticatorClientPinResponse {
+            key_agreement: None,
+            pin_token: None,
+            retries: Some(persistent_store.pin_retries() as u64),
+        })
+    }
+
+    fn process_get_key_agreement(&self) -> Result<AuthenticatorClientPinResponse, Ctap2StatusCode> {
+        let pk = self.key_agreement_key.genpk();
+        Ok(AuthenticatorClientPinResponse {
+            key_agreement: Some(CoseKey::from(pk)),
+            pin_token: None,
+            retries: None,
+        })
+    }
+
+    fn process_set_pin(
+        &mut self,
+        persistent_store: &mut PersistentStore,
+        key_agreement: CoseKey,
+        pin_auth: Vec<u8>,
+        new_pin_enc: Vec<u8>,
+    ) -> Result<(), Ctap2StatusCode> {
+        if persistent_store.pin_hash().is_some() {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID);
+        }
+        let pk: crypto::ecdh::PubKey = CoseKey::try_into(key_agreement)?;
+        let shared_secret = self.key_agreement_key.exchange_x_sha256(&pk);
+
+        if !check_pin_auth(&shared_secret, &new_pin_enc, &pin_auth) {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID);
+        }
+
+        let aes_enc_key = crypto::aes256::EncryptionKey::new(&shared_secret);
+        let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+        if !check_and_store_new_pin(persistent_store, &aes_dec_key, new_pin_enc) {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_POLICY_VIOLATION);
+        }
+        persistent_store.reset_pin_retries();
+        Ok(())
+    }
+
+    fn process_change_pin(
+        &mut self,
+        rng: &mut impl Rng256,
+        persistent_store: &mut PersistentStore,
+        key_agreement: CoseKey,
+        pin_auth: Vec<u8>,
+        new_pin_enc: Vec<u8>,
+        pin_hash_enc: Vec<u8>,
+    ) -> Result<(), Ctap2StatusCode> {
+        if persistent_store.pin_retries() == 0 {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_BLOCKED);
+        }
+        let pk: crypto::ecdh::PubKey = CoseKey::try_into(key_agreement)?;
+        let shared_secret = self.key_agreement_key.exchange_x_sha256(&pk);
+
+        let mut auth_param_data = new_pin_enc.clone();
+        auth_param_data.extend(&pin_hash_enc);
+        if !check_pin_auth(&shared_secret, &auth_param_data, &pin_auth) {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID);
+        }
+
+        let aes_enc_key = crypto::aes256::EncryptionKey::new(&shared_secret);
+        let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+        self.check_pin_hash_enc(rng, persistent_store, &aes_dec_key, pin_hash_enc)?;
+
+        if !check_and_store_new_pin(persistent_store, &aes_dec_key, new_pin_enc) {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_POLICY_VIOLATION);
+        }
```

Comment:
This code has a lot of overlap with the previous function. It would be worth factoring it into a utility function to avoid code duplication.

Reply:
It's quite entangled. The lower of `change_pin` needs the `self.check_pin_hash_enc` between the shared code. The `shared_secret` is used multiple times in the function, too. I could have a function that does

`        let pk: crypto::ecdh::PubKey = CoseKey::try_into(key_agreement)?;`
`        let shared_secret = self.key_agreement_key.exchange_x_sha256(&pk);`
`        if !check_pin_auth(&shared_secret, &message, &pin_auth) {`
`            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID);`
`        }`
`        return shared_secret;`

But I don't like the side effect of generating shared_secret and passing it around. What was your idea?